idseq_dag/steps/run_validate_input.py

```diff
@@ -0,0 +1,207 @@
+
+                if is_fastq:
+                    identifier2_l = input_f.readline()
+                    if len(identifier2_l) == 0:
+                        raise RuntimeError(f"Invalid FASTQ file, unexpected EOF: {file}")
+
+                    quality_l = input_f.readline()
+                    if len(quality_l) == 0:
+                        raise RuntimeError(f"Invalid FASTQ file, unexpected EOF: {file}")
+
+                if is_fastq:
+                    if identifier_l[0] != '@' or identifier2_l[0] != '+':
+                        # may be FASTQ file with multi-line reads, requires full check
+                        return False
+                else:
+                    if identifier_l[0] != '>':
+                         # may be FASTA file with multi-line reads, requires full check
+                        return False
+
+                if fragment_length == 0:
+                    fragment_length = len(identifier_l)
+                    if fragment_length < 50 or fragment_length > 500:
+                         # non-standard fragment lengths require more detailed examination
+                        return False
+
+                if fragment_length != len(identifier_l) or fragment_length != len(identifier2_l):
+                     # file does not meet "quick check" requirements since fragments/quality
+                     # scores are not all of same length
+                    return False
+
+        return True
+
+    def truncate_file(self, infile, outfile, is_fastq, max_fragments):
+        if is_fastq:
+            num_lines = max_fragments * 4
+        else:
+            num_lines = max_fragments * 2
+        command.execute(f"head -n {num_lines} {infile} > {outfile}")
+        self.summary_dict['50-500'] = count.reads_in_group(self.output_files_local()[0:2])
+        return
+
+    # full_check_and_truncate_file does an exhaustive check of the input file, up to
+    # max_fragments, and reformats the output to conform to what the rest of the 
+    # computational pipeline expects (single-line reads of max-length 10,000). After
+    # viewing max_fragments reads or encountering EOF, the function returns.
+    #
+    # Throws an exception in the case of an unrecoverable abnormality
+    def full_check_and_truncate_file(self, infile, outfile, is_fastq, max_fragments):
+        num_fragments = 0
+        fragment_length = 0
+
+        with open(infile, 'r', encoding='utf-8') as input_f, open(outfile, 'w') as output_f:
+            next_line = input_f.readline()
+            while True:
+                num_fragments += 1
+                if num_fragments > max_fragments:
+                    break
+
+                identifier_l = next_line
+                if len(identifier_l) == 0: # EOF
+                    break
+
+                read_l = input_f.readline()
+                if len(read_l) == 0:
+                    raise RuntimeError(f"Invalid input file, unexpected EOF: {file}")
```

Comment:
I think it would be a great opportunity to put these validation error messages into a dedicated output file as well. From there, it will just be a tiny change to show the message to the user -- which will save us tons of on-call work.

Reply:
neat idea - i think maybe i'll put it in the same file i was using for the summary counts and store it in json

Reply:
@cdebourcy ready for another review